xcode/Subconscious/Shared/Components/AddressBook/AddFriendView.swift

```diff
@@ -7,24 +7,53 @@
 
 import SwiftUI
 import ObservableStore
+import CodeScanner
+
+typealias Validator<I, O> = (I) -> O?
+
+struct FormField<I, O> {
+    var value: I
+    var validate: Validator<I, O>
+    var touched: Bool = false
+    
+    var validated: O? {
+        get {
+            validate(value)
+        }
+    }
+    var isValid: Bool {
+        get {
+            validated != nil
+        }
+    }
+    var hasError: Bool {
+        get {
+            validated == nil && touched
+        }
+    }
+}
```

Comment:
Not trying to venture too far into an form framework here, this seems like enough for ephemeral forms with a few fields. There's much more to explore RE: validation and error messages later.